File: trunk/CCMContainer/ccmcommon/c++/src/Cookie_impl.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


// Std Files
#include <string>
#include <sstream>
#include <iostream>

// Cdmw Files
#include <Foundation/orbsupport/CORBA.hpp>
#include <Foundation/orbsupport/ExceptionMinorCodes.hpp>

#include <Foundation/ossupport/OS.hpp>

#include <CCMContainer/ccmcommon/Cookie_impl.hpp>
#include <CCMContainer/ccmcommon/CCMUtils.hpp>


namespace Cdmw {

namespace CCM {
    
namespace Common {
    //
    // IDL:omg.org/Components/Cookie:1.0
    // 

    // Constructor
    Cookie_impl::Cookie_impl()
    {
    }

    Cookie_impl::Cookie_impl (const std::string & s)
    {
        PRINT_INFO("Cookie created : s = " << s);
        CORBA::OctetSeq_var cookie_value = string_to_OctetSeq(s.c_str());

        // set cookieValue
        cookieValue(cookie_value.in());
    }

    Cookie_impl::Cookie_impl (const CORBA::OctetSeq & seq)
        : OBV_Components::Cookie()
    {
        cookieValue( seq );
        CORBA::String_var s = OctetSeq_to_string(cookieValue());
        PRINT_INFO("Cookie created : s = " << s.in());
    }

    Cookie_impl* Cookie_impl::create_cookie()
        throw(Components::ExceededConnectionLimit)
    {
        try {
            unsigned long counter = create_identifier();
        
            using namespace Cdmw::OsSupport;
            OS::Timeval tv = OS::get_time();
        
        
            std::ostringstream ostr;
            ostr << tv.seconds << ":" << tv.microseconds << ":" << counter;
            
            return new Cookie_impl(ostr.str());
        } catch (...) {
            throw Components::ExceededConnectionLimit();
        }
    }

    // Copy Constructor
    Cookie_impl::Cookie_impl (const Cookie_impl& right)
    {
        cookieValue(right.cookieValue());
    }

    // Destructor
    Cookie_impl::~Cookie_impl ()
    {
    }

    std::string  
    Cookie_impl::get_cookie_string() const
    {
        CORBA::String_var str = OctetSeq_to_string(cookieValue());
        std::string std_str(str.in());
        return std_str;
    }


    CORBA::ValueBase* 
    Cookie_impl::_copy_value()
    {
        return new Cookie_impl(cookieValue());
    }

    
    bool 
    Cookie_impl::operator<(const Cookie_impl& right) const
        throw()
    {
        CORBA::String_var lv = OctetSeq_to_string(cookieValue());
        CORBA::String_var rv = OctetSeq_to_string(right.cookieValue());
            
        std::string left_value(lv.in());
        std::string right_value(rv.in());

        return (left_value < right_value);        
    }

    bool 
    Cookie_impl::operator==(const Cookie_impl& right) const
        throw()
    {
        CORBA::String_var lv = OctetSeq_to_string(cookieValue());
        CORBA::String_var rv = OctetSeq_to_string(right.cookieValue());;
            
        std::string left_value(lv.in());
        std::string right_value(rv.in());

        return (left_value == right_value);        
    }



    /******************************/
    /*         Factory            */
    /******************************/


    CookieFactory::CookieFactory()
    {
    }
    
    CookieFactory::~CookieFactory()
    {
    }

    CORBA::ValueBase * 
    CookieFactory::create_for_unmarshal ()
    {
        return new Cookie_impl();
    }

} // End namespace CIF
} // End namespace CCM
} // End namespace Cdmw

File: tags/CARDAMOM_V1_1_CE/CCMContainer/ccmcommon/java/src/cdmw/ccm/common/CookieImpl.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.ccm.common;

import cdmw.ossupport.OS;
import cdmw.ossupport.Timeval;

/**
 * Implementation of the org.omg.Components.Cookie valuetype.
 * It redefines the equals() method and the hashCode() method,
 * allowing use of Cookies in a Map.
 */
public class CookieImpl extends org.omg.Components.Cookie 
    implements Comparable {

    /**
     * Default Constructor.
     * cookieValue is set to null.
     */
    public CookieImpl() {
        this.cookieValue = null;
    }

    /**
     * Constructor.
     *
     * @param cookieValue the value to be set for the new Cookie
     */
    public CookieImpl(byte[] cookieValue) {
        this.cookieValue = cookieValue;
    }
    
    /**
     * Constructor.
     *
     * @param cookieValue the value to be set for the new Cookie
     */
    public CookieImpl(String cookieValue) {
        this.cookieValue = cookieValue.getBytes();
    }
    
    /**
     * Returns the cookie value as a String object.
     * 
     * @return The cookie value.
     */
    public String getCookieString() {
        return new String(this.cookieValue);
    }

    /**
     * Compares this object to the specified object. 
     * The result is true if and only if the argument is not null 
     * and is a CookieImpl object that contains the same cookieValue 
     * as this object.
     *
     * @param the object to compare with.
     *
     * @return true if the objects are the same; false otherwise.
     */
    public boolean equals(Object obj) {

        // if not the same type, not equals
        if (!(obj instanceof CookieImpl)) {
            return false;
        }

        // get obj.cookieValue
        byte[] b = ((CookieImpl) obj).cookieValue;

        // if cookieValues have different size, not equals
        if (b.length != cookieValue.length) {
            return false;
        }

        // if a byte differs, not equals
        for (int i = 0; i < b.length; i++) {
            if (b[i] != cookieValue[i]) {
                return false;
            }
        }

        // they are equals
        return true;
    }

    /**
     * Returns a hashcode for this CookieImpl.
     *
     * @return a hash code value for this object.
     */
    public int hashCode() {
        int hash = 0;

        // for each byte of the cookieValue
        for (int i = 0;(i < cookieValue.length); i++) {

            // create a corresponding mask (a byte has 8 bits)
            // (shift is made with modulo sizeof(int)) 
            int mask = cookieValue[i] << (8 * i);

            // apply mask to hash code with xor
            // (xor reduce probability to have only 1 bits 
            // with large cookieValues)
            hash = hash ^ mask;
        }

        return hash;
    }
    
    /**
     * Compares this object with the specified object for order.
     * 
     * @param o the Object to be compared.
     * @return a negative integer, zero, or a positive integer as this object 
     * is less than, equal to, or greater than the specified object.
     * @throws ClassCastException if the specified object's type prevents it 
     * from being compared to this Object
     */
    public int compareTo(Object o) {
        CookieImpl other = (CookieImpl) o;
        return getCookieString().compareTo(other.getCookieString());
    }

    /**
     * Static method to create a Cookie.
     * 
     * @return CookieImpl
     * @throws ExceededConnectionLimit
     */
    public static org.omg.Components.Cookie createCookie()
        throws org.omg.Components.ExceededConnectionLimit {

        try {
            int counter = CCMUtils.createIdentifier();
            Timeval tv = OS.getTime();
            String ostr = new String(tv.getSeconds() + ":" + 
                tv.getMicroseconds() + ":" + counter);
            return new CookieImpl(ostr);
        } catch (Exception e) {
            throw new org.omg.Components.ExceededConnectionLimit();
        }

    }

}

